File: YoloWebCam/webcamworker.cpp
#include "webcamworker.h"
#include <QThread>

WebcamWorker::WebcamWorker(QObject *parent)
    : QObject(parent), running(false)
{
}

WebcamWorker::~WebcamWorker()
{
    stop();
}

void WebcamWorker::start()
{
    if (running) return;

    running = true;
    cap.open(0); // 0번 카메라 열기

    if (!cap.isOpened()) {
        qWarning("Failed to open webcam.");
        running = false;
        return;
    }

    while (running) {
        cv::Mat frame;
        cap >> frame;
        if (frame.empty())
            continue;

        cv::cvtColor(frame, frame, cv::COLOR_BGR2RGB);
        QImage image(frame.data, frame.cols, frame.rows, frame.step, QImage::Format_RGB888);

        emit frameReady(image.copy()); // QImage 복사해서 보내기

        QThread::msleep(30); // 대략 30FPS
    }

    cap.release();
}

void WebcamWorker::stop()
{
    QMutexLocker locker(&mutex);
    running = false;
}

File: YoloPyQT/webcam_worker.py
import cv2
import sys
from PySide6.QtCore import Signal, QObject
from PySide6.QtGui import QImage

class WebcamWorker(QObject):
    frameCaptured = Signal(QImage)

    def __init__(self):
        super().__init__()
        self.running = False
        self.cap = None

    def start(self):
        sys.stdout.flush()
        self.cap = cv2.VideoCapture(0)
        self.running = True

        if not self.cap.isOpened():
            print("[ERROR] Webcam could not be opened.")
            return False

        while self.running:
            ret, frame = self.cap.read()
            if not ret:
                continue
            rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
            h, w, ch = rgb.shape
            bytes_per_line = ch * w
            qimg = QImage(rgb.data, w, h, bytes_per_line, QImage.Format_RGB888)
            self.frameCaptured.emit(qimg)

    def stop(self):
        self.running = False
        if self.cap and self.cap.isOpened():
            self.cap.release()


